{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "tcbBdgYoMf8O",
    "outputId": "c7fc0a25-9760-42d6-8768-e17b560cbdee",
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "# import umap\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pooch\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scvi\n",
    "import torch\n",
    "from matplotlib import rcParams\n",
    "rcParams['pdf.fonttype'] = 42 # enables correct plotting of text\n",
    "rcParams['figure.figsize'] = (12,12)\n",
    "\n",
    "import muon as mu\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import snapatac2 as snap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:63: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:70: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:71: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:80: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n",
      "/tmp/ipykernel_15410/1486545137.py:129: RuntimeWarning: divide by zero encountered in log10\n",
      "  weighted_score = log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:136: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:137: RuntimeWarning: divide by zero encountered in log10\n",
      "  gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
      "/tmp/ipykernel_15410/1486545137.py:146: RuntimeWarning: invalid value encountered in scalar divide\n",
      "  data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
      "/home/macera/.conda/envs/scanpy_basic/lib/python3.9/site-packages/scipy/stats/_stats_py.py:9690: RuntimeWarning: divide by zero encountered in log\n",
      "  statistic = -2 * np.sum(np.log(pvalues))\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import snapatac2 as snap\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import os\n",
    "from scipy.stats import combine_pvalues\n",
    "\n",
    "def parse_index(index):\n",
    "    chrom, positions = index.split(':')\n",
    "    start, end = positions.split('-')\n",
    "    return chrom, int(start), int(end)\n",
    "\n",
    "def csv_to_bed(csv_file, bed_file):\n",
    "    # Read the CSV file\n",
    "    df = pd.read_csv(csv_file, index_col=0)\n",
    "    \n",
    "    # Parse the index to extract chromosome, start, and end positions\n",
    "    parsed_data = [parse_index(idx) for idx in df.index]\n",
    "    bed_df = pd.DataFrame(parsed_data, columns=['chromosome', 'start', 'end'])\n",
    "    \n",
    "    # Save to BED format (tab-separated)\n",
    "    bed_df.to_csv(bed_file, sep='\\t', header=False, index=False)\n",
    "    print(f\"Converted {csv_file} to {bed_file}\")\n",
    "\n",
    "for l in ['l1', 'l2']:\n",
    "    for file in os.listdir(f\"DAR/csv/{l}/\"):\n",
    "        if file.endswith('.csv') and file.startswith('DAR'):\n",
    "            input_csv = f\"DAR/csv/{l}/{file}\"\n",
    "\n",
    "            # Save the ranked list to a CSV file\n",
    "            file_save = file.replace('DAR', 'DAG_distal')\n",
    "            output_csv = f\"DAR/csv/{l}/DAG/{file_save}\"\n",
    "\n",
    "            df = pd.read_csv(input_csv)\n",
    "            if not os.path.exists(output_csv):\n",
    "                \n",
    "                # Initialize the network using your existing setup\n",
    "                network = snap.tl.init_network_from_annotation(\n",
    "                    regions=df['Feature_name'].tolist(),\n",
    "                    anno_file=snap.genome.GRCh38,\n",
    "                    # upstream=5000,\n",
    "                    # downstream=100,\n",
    "                    # only_coding=True\n",
    "                )\n",
    "\n",
    "                # Initialize a dictionary to store gene scores\n",
    "                gene_scores = defaultdict(lambda: {\n",
    "                    \"score_sum\": 0, \"region_count\": 0,\n",
    "                    \"lfc_sum\": 0, \"lfc_weight_sum\": 0,\n",
    "                    \"pval_list\": []\n",
    "                })\n",
    "\n",
    "                for edge in network.edge_list():\n",
    "                    region = network[edge[0]].id\n",
    "                    gene = network[edge[1]].id\n",
    "                    \n",
    "                    if region in df['Feature_name'].values:\n",
    "                        region_data = df[df['Feature_name'] == region]\n",
    "                        log2_fold_change = region_data['log2(fold_change)'].values[0]\n",
    "                        p_value = region_data['p-value'].values[0]\n",
    "                        if p_value < 1e-50:\n",
    "                            p_value = 1e-50 \n",
    "                        \n",
    "                        # Calculate weighted score for each region\n",
    "                        weighted_score = log2_fold_change * (-np.log10(p_value))\n",
    "                        \n",
    "                        # Accumulate scores\n",
    "                        gene_scores[gene][\"score_sum\"] += weighted_score\n",
    "                        gene_scores[gene][\"region_count\"] += 1\n",
    "                        \n",
    "                        # Accumulate weighted log2 fold changes\n",
    "                        gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
    "                        gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
    "                        \n",
    "                        # Collect p-values for later combination\n",
    "                        gene_scores[gene][\"pval_list\"].append(p_value)\n",
    "\n",
    "                # Process the gene_scores to aggregate LFC and p-values\n",
    "                for gene, data in gene_scores.items():\n",
    "                    # Compute the average LFC, weighted by the significance of each region\n",
    "                    if data[\"lfc_weight_sum\"] != 0:\n",
    "                        data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
    "                    else:\n",
    "                        data[\"average_lfc\"] = 0\n",
    "                    \n",
    "                    # Combine p-values using Fisher's method\n",
    "                    if len(data[\"pval_list\"]) > 1:\n",
    "                        _, combined_pval = combine_pvalues(data[\"pval_list\"], method='fisher')\n",
    "                        data[\"combined_pval\"] = combined_pval\n",
    "                    else:\n",
    "                        data[\"combined_pval\"] = data[\"pval_list\"][0] if data[\"pval_list\"] else 1\n",
    "\n",
    "                # Convert the dictionary to a DataFrame for ranking\n",
    "                gene_df = pd.DataFrame.from_dict(gene_scores, orient='index')\n",
    "                gene_df['average_score'] = gene_df['score_sum'] / gene_df['region_count']\n",
    "                gene_df = gene_df.sort_values(by=\"average_score\", ascending=False)\n",
    "\n",
    "                gene_df.to_csv(output_csv)\n",
    "\n",
    "            # Processing overlap file (as in DAG_overlap)\n",
    "            file_save = file.replace('DAR', 'DAG_overlap')\n",
    "            output_csv = f\"DAR/csv/{l}/DAG/{file_save}\"\n",
    "\n",
    "            if not os.path.exists(output_csv):\n",
    "                \n",
    "                network = snap.tl.init_network_from_annotation(\n",
    "                    regions=df['Feature_name'].tolist(),\n",
    "                    anno_file=snap.genome.GRCh38,\n",
    "                    upstream=5000,\n",
    "                    downstream=100,\n",
    "                    # only_coding=True\n",
    "                )\n",
    "\n",
    "                # Initialize a dictionary to store gene scores\n",
    "                gene_scores = defaultdict(lambda: {\n",
    "                    \"score_sum\": 0, \"region_count\": 0,\n",
    "                    \"lfc_sum\": 0, \"lfc_weight_sum\": 0,\n",
    "                    \"pval_list\": []\n",
    "                })\n",
    "\n",
    "                for edge in network.edge_list():\n",
    "                    region = network[edge[0]].id\n",
    "                    gene = network[edge[1]].id\n",
    "                    \n",
    "                    if region in df['Feature_name'].values:\n",
    "                        region_data = df[df['Feature_name'] == region]\n",
    "                        log2_fold_change = region_data['log2(fold_change)'].values[0]\n",
    "                        p_value = region_data['p-value'].values[0]\n",
    "                        if p_value < 1e-50:\n",
    "                            p_value = 1e-50 \n",
    "                        \n",
    "                        # Calculate weighted score for each region\n",
    "                        weighted_score = log2_fold_change * (-np.log10(p_value))\n",
    "                        \n",
    "                        # Accumulate scores\n",
    "                        gene_scores[gene][\"score_sum\"] += weighted_score\n",
    "                        gene_scores[gene][\"region_count\"] += 1\n",
    "                        \n",
    "                        # Accumulate weighted log2 fold changes\n",
    "                        gene_scores[gene][\"lfc_sum\"] += log2_fold_change * (-np.log10(p_value))\n",
    "                        gene_scores[gene][\"lfc_weight_sum\"] += (-np.log10(p_value))\n",
    "                        \n",
    "                        # Collect p-values for later combination\n",
    "                        gene_scores[gene][\"pval_list\"].append(p_value)\n",
    "\n",
    "                # Process the gene_scores to aggregate LFC and p-values\n",
    "                for gene, data in gene_scores.items():\n",
    "                    # Compute the average LFC, weighted by the significance of each region\n",
    "                    if data[\"lfc_weight_sum\"] != 0:\n",
    "                        data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"lfc_weight_sum\"]\n",
    "                    else:\n",
    "                        data[\"average_lfc\"] = 0\n",
    "                    \n",
    "                    # Combine p-values using Fisher's method\n",
    "                    if len(data[\"pval_list\"]) > 1:\n",
    "                        _, combined_pval = combine_pvalues(data[\"pval_list\"], method='fisher')\n",
    "                        data[\"combined_pval\"] = combined_pval\n",
    "                    else:\n",
    "                        data[\"combined_pval\"] = data[\"pval_list\"][0] if data[\"pval_list\"] else 1\n",
    "\n",
    "                # Convert the dictionary to a DataFrame for ranking\n",
    "                gene_df = pd.DataFrame.from_dict(gene_scores, orient='index')\n",
    "                gene_df['average_score'] = gene_df['score_sum'] / gene_df['region_count']\n",
    "                gene_df = gene_df.sort_values(by=\"average_score\", ascending=False)\n",
    "\n",
    "                gene_df.to_csv(output_csv)"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "name": "PeakVI_tutorial.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "scanpy_basic",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
