{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import rcParams\n",
    "rcParams['pdf.fonttype'] = 42 # enables correct plotting of text\n",
    "rcParams['figure.figsize'] = (12,12)\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.11.4'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import scipy\n",
    "scipy.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def csv_to_bed_df(csv_file, bed_file):\n",
    "    # Read the CSV file\n",
    "    df = pd.read_csv(csv_file,sep='\\t', index_col=None)\n",
    "    print(df)\n",
    "    bed_df = df[['chrom', 'start', 'end']]\n",
    "    bed_df.rename(columns = ['chromosome', 'start', 'end'], inplace=True)\n",
    "    # Parse the index to extract chromosome, start, and end positions\n",
    "    # parsed_data = [parse_index(idx) for idx in df.index]\n",
    "    # bed_df = pd.DataFrame(df, columns=['chromosome', 'start', 'end'])\n",
    "    \n",
    "    # Save to BED format (tab-separated)\n",
    "    bed_df.to_csv(bed_file, sep='\\t', header=False, index=False)\n",
    "    print(f\"Converted {csv_file} to {bed_file}\")\n",
    "def parse_index(index):\n",
    "    chrom, positions = index.split(':')\n",
    "    start, end = positions.split('-')\n",
    "    return chrom, int(start), int(end)\n",
    "def index_to_bed(index, bed_file):\n",
    "    # Read the CSV file\n",
    "\n",
    "    # Parse the index to extract chromosome, start, and end positions\n",
    "    parsed_data = [parse_index(idx) for idx in index]\n",
    "    bed_df = pd.DataFrame(parsed_data, columns=['chromosome', 'start', 'end'])\n",
    "    \n",
    "    # Save to BED format (tab-separated)\n",
    "    bed_df.to_csv(bed_file, sep='\\t', header=False, index=False)\n",
    "    print(f\"Converted {csv_file} to {bed_file}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import snapatac2 as snap\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "def parse_index(index):\n",
    "    chrom, positions = index.split(':')\n",
    "    start, end = positions.split('-')\n",
    "    return chrom, int(start), int(end)\n",
    "\n",
    "def csv_to_bed(csv_file, bed_file):\n",
    "    # Read the CSV file\n",
    "    df = pd.read_csv(csv_file, index_col=0)\n",
    "    \n",
    "    # Parse the index to extract chromosome, start, and end positions\n",
    "    parsed_data = [parse_index(idx) for idx in df.index]\n",
    "    bed_df = pd.DataFrame(parsed_data, columns=['chromosome', 'start', 'end'])\n",
    "    \n",
    "    # Save to BED format (tab-separated)\n",
    "    bed_df.to_csv(bed_file, sep='\\t', header=False, index=False)\n",
    "    print(f\"Converted {csv_file} to {bed_file}\")\n",
    "\n",
    "for l in ['l1', 'l2']:\n",
    "    for file in os.listdir(f\"DAR/csv/{l}/\"):\n",
    "        if file.endswith('.csv') and file.startswith('DAR'):\n",
    "            input_csv = f\"DAR/csv/{l}/{file}\"\n",
    "\n",
    "            # Processing overlap file (as in DAG_overlap)\n",
    "            file_save = file.replace('DAR', 'DAG_Overlap')\n",
    "            output_csv = f\"DAR/csv/{l}/DAG/{file_save}\"\n",
    "\n",
    "            df = pd.read_csv(input_csv)\n",
    "\n",
    "            if not os.path.exists(output_csv):\n",
    "                \n",
    "\n",
    "                from collections import defaultdict\n",
    "\n",
    "                from scipy.stats import norm\n",
    "                network = snap.tl.init_network_from_annotation(\n",
    "                    regions=df['Feature_name'].tolist(),\n",
    "                    anno_file=snap.genome.GRCh38,\n",
    "                    upstream=2000,\n",
    "                    downstream=500,\n",
    "                    # only_coding=True\n",
    "                )\n",
    "                # Initialize the gene score dictionary\n",
    "                gene_scores = defaultdict(lambda: {\n",
    "                    \"lfc_sum\": 0, \"region_count\": 0, \"z_sum\": 0,\"score_sum\":0\n",
    "                })\n",
    "\n",
    "                for edge in network.edge_list():\n",
    "                    region = network[edge[0]].id\n",
    "                    gene = network[edge[1]].id\n",
    "\n",
    "                    if region in df['Feature_name'].values:\n",
    "                        region_data = df[df['Feature_name'] == region]\n",
    "                        log2_fold_change = region_data['log2(fold_change)'].values[0]\n",
    "                        p_value = region_data['p-value'].values[0]\n",
    "                        if p_value < 1e-50:  # Cap extremely low p-values to avoid numerical issues\n",
    "                            p_value = 1e-50 \n",
    "\n",
    "                        # Convert p-value to Z-score for Stouffer's method\n",
    "                        z_score = norm.ppf(1 - p_value)\n",
    "\n",
    "                        weighted_score = log2_fold_change * (-np.log10(p_value))\n",
    "                        \n",
    "                        # Accumulate scores\n",
    "                        gene_scores[gene][\"score_sum\"] += weighted_score\n",
    "                        # Accumulate scores\n",
    "                        gene_scores[gene][\"lfc_sum\"] += log2_fold_change\n",
    "                        gene_scores[gene][\"z_sum\"] += z_score\n",
    "                        gene_scores[gene][\"region_count\"] += 1\n",
    "\n",
    "                # Process the gene_scores to aggregate LFC and p-values\n",
    "                for gene, data in gene_scores.items():\n",
    "                    # Compute the average LFC as an unweighted mean\n",
    "                    if data[\"region_count\"] > 0:\n",
    "                        data[\"average_lfc\"] = data[\"lfc_sum\"] / data[\"region_count\"]\n",
    "                    else:\n",
    "                        data[\"average_lfc\"] = 0\n",
    "\n",
    "                    # Combine p-values using Stouffer's method\n",
    "                    if data[\"region_count\"] > 1:\n",
    "                        combined_z = data[\"z_sum\"] / np.sqrt(data[\"region_count\"])\n",
    "                        data[\"combined_pval\"] = 1 - norm.cdf(combined_z)\n",
    "                    else:\n",
    "                        # If there's only one p-value, use it directly\n",
    "                        data[\"combined_pval\"] = 1 - norm.cdf(data[\"z_sum\"]) if data[\"region_count\"] == 1 else 1\n",
    "\n",
    "                # Convert gene_scores to a DataFrame for easier analysis\n",
    "                gene_scores_df = pd.DataFrame.from_dict(gene_scores, orient='index')\n",
    "                gene_scores_df['average_score'] = gene_scores_df['score_sum'] / gene_scores_df['region_count']\n",
    "\n",
    "                gene_scores_df.reset_index(inplace=True)\n",
    "                gene_scores_df.rename(columns={'index': 'Gene'}, inplace=True)\n",
    "\n",
    "                gene_scores_df.to_csv(output_csv)\n"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "name": "PeakVI_tutorial.ipynb",
   "provenance": [],
   "toc_visible": true
  },
  "kernelspec": {
   "display_name": "scanpy_basic",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
